File: android/app/src/main/java/app/maptalk/data/LocalMediaStore.kt
package app.maptalk.data

import android.content.Context
import java.io.File
import java.util.UUID

/** Files for local-demo image messages under the app's files directory. */
class LocalMediaStore(context: Context) {

    private val root: File = File(context.filesDir, "maptalk-media").also { it.mkdirs() }

    fun save(jpeg: ByteArray, preferredName: String? = null): String {
        val name = preferredName ?: "${UUID.randomUUID()}.jpg"
        val file = File(root, name)
        file.parentFile?.mkdirs()
        file.writeBytes(jpeg)
        return name
    }

    fun delete(relativePath: String) {
        File(root, relativePath).delete()
    }

    fun saveAudio(bytes: ByteArray, ext: String = "m4a"): String {
        val name = "${UUID.randomUUID()}.$ext"
        File(root, name).writeBytes(bytes)
        return name
    }

    fun saveVideo(bytes: ByteArray, ext: String = "mp4"): String {
        val name = "${UUID.randomUUID()}.$ext"
        File(root, name).writeBytes(bytes)
        return name
    }

    fun file(relativePath: String): File = File(root, relativePath)
}

File: ios/MapTalk/Data/LocalMediaStore.swift
import Foundation

/// Files for local-demo image messages. Lives under Application Support so iTunes backups can
/// keep them, and so we never depend on a network path while trying the app on a phone.
enum LocalMediaStore {

    private static let folderName = "maptalk-media"

    static var root: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let dir = base.appendingPathComponent(folderName, isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    /// Writes JPEG bytes and returns a relative path (the message stores this, not an absolute URL).
    static func save(jpeg: Data, preferredName: String? = nil) throws -> String {
        let name = preferredName ?? "\(UUID().uuidString).jpg"
        let file = root.appendingPathComponent(name)
        try jpeg.write(to: file, options: .atomic)
        return name
    }

    static func save(audio: Data, ext: String = "m4a") throws -> String {
        let name = "\(UUID().uuidString).\(ext)"
        let file = root.appendingPathComponent(name)
        try audio.write(to: file, options: .atomic)
        return name
    }

    static func save(video: Data, ext: String = "mp4") throws -> String {
        let name = "\(UUID().uuidString).\(ext)"
        let file = root.appendingPathComponent(name)
        try video.write(to: file, options: .atomic)
        return name
    }

    static func url(forRelativePath path: String) -> URL {
        root.appendingPathComponent(path)
    }

    static func data(forRelativePath path: String) -> Data? {
        try? Data(contentsOf: url(forRelativePath: path))
    }
}
